src/app/clusters/door-lock-server/door-lock-server.cpp

```diff
@@ -140,30 +143,115 @@ bool emberAfDoorLockClusterLockDoorCallback(chip::app::CommandHandler * commandO
                                             const chip::app::ConcreteCommandPath & commandPath,
                                             const chip::app::Clusters::DoorLock::Commands::LockDoor::DecodableType & commandData)
 {
-    emberAfDoorLockClusterPrintln("Received Lock Door command (not implemented)");
+    emberAfDoorLockClusterPrintln("Received Lock Door command");
+    bool success = false;
 
-    // TODO: Implement door locking by calling emberAfPluginDoorLockOnDoorLockCommand
+    chip::EndpointId endpoint = commandPath.mEndpointId;
 
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
-    return true;
+    bool require_pin = false;
+    Attributes::RequirePINforRemoteOperation::Get(endpoint, &require_pin);
+
+    if(require_pin)
+    {
+        if(!commandData.pinCode.HasValue())
+        {
+            success = false; // Just to be explicit. success == false at this point anywyay
+        }
+        else
+        {
+            const char *cmd_pin = reinterpret_cast<const char*>(commandData.pinCode.Value().data());
+            char *mPin = DoorLockServer::Instance().mPin;
+            if(strncmp(mPin, cmd_pin, 4) == 0)
+            {
+                success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, cmd_pin);
+            }
+            else
+            {
+                success = false;
+            }
+        }
+    }
+    else
+    {
+        const char *cmd_pin = commandData.pinCode.HasValue() ?
+            reinterpret_cast<const char*>(commandData.pinCode.Value().data()) :
+            NULL;
+        success = emberAfPluginDoorLockOnDoorLockCommand(endpoint, cmd_pin);
+    }
+
+    emberAfSendImmediateDefaultResponse(success ? EMBER_ZCL_STATUS_SUCCESS : EMBER_ZCL_STATUS_FAILURE);
+
+    // Should be called after reponse sent (indicating command complete) since
+    // the attribute Set call in SetLockState will error out if the command is
+    // still pending
```

Comment:
Wait, why will the attribute set call error out?

Reply:
I was getting the following error from the app output when calling Set (Note: Lines numbers may not match anymore):
[1639081594.128290][29991:29991] CHIP:-: ../../examples/lock-app/linux/third_party/connectedhomeip/src/app/CommandHandler.cpp:389: CHIP Error 0x00000003: Incorrect state at ../../examples/lock-app/linux/third_party/connectedhomeip/src/app/CommandHandler.cpp:308
[1639081594.128345][29991:29991] CHIP:ZCL: Failed to dispatch command, TLVError=../../examples/lock-app/linux/third_party/connectedhomeip/zzz_generated/app-common/app-common/zap-generated/cluster-objects.cpp:9015: Success

The actual line in CommandHandler that was throwing the error was:
// We must not be in the middle of preparing a command, or having prepared or sent one.
VerifyOrReturnError(mState == State::Idle, CHIP_ERROR_INCORRECT_STATE);

The error went away when I moved the Set call after the emberAfSendImmediateDefaultResponse() call.

Reply:
I just tried putting the Set before the send, and it was working fine.  Please describe the exact steps to reproduce that show this problem?  Exact commands executed in which exact client, etc.

Reply:
Well I can't seem to reproduce the issue. I tried building like you did (for onnetwork pairing) as well as using two Pis and running ble-wifi pairing. I was working off an older master when I saw the issue so perhaps it's been fixed since.

Reply:
In that case, it's probably better to update the internal state before sending the response.